video/out/vo.c

```diff
@@ -1060,6 +1060,30 @@ static struct mp_image *get_image_vo(void *ctx, int imgfmt, int w, int h,
     return vo->driver->get_image(vo, imgfmt, w, h, stride_align);
 }
 
+void vo_drop_unrendered_frame(struct vo *vo)
+{
+    struct vo_internal *in = vo->in;
+
+    pthread_mutex_lock(&in->lock);
+
+    if (!in->frame_queued)
+        goto end;
+
+    if ((in->frame_queued->pts + in->frame_queued->duration) > mp_time_us())
+        goto end;
+
+    MP_VERBOSE(vo, "Dropping unrendered frame (pts %li)\n", in->frame_queued->pts);
+
+    talloc_free(in->frame_queued);
+    in->frame_queued = NULL;
+    in->hasframe = false;
+    pthread_cond_broadcast(&in->wakeup);
+    wakeup_core(vo);
+
+end:
+    pthread_mutex_unlock(&in->lock);
+}
+
```

Comment:
This function is still Bad. Because of how display sync is done, because audio timing is a different thing altogether, and due to the fact that computing the timing is still in the player core, vo_frame (the thing `in->frame_queued` points to) might be more complex than you might think. For example, the pts field may have no meaning (if display sync is active), and a frame can be repeated multiple times (if `num_vsyncs != 1`, so you'd actually drop multiple frames).

Also I'm not really sure if hasframe is properly handled. This field basically determines whether the VO user (the player core) has queued a frame, which it needs for... various reasons.

Or the way the frame is discarded. You will notice that frame_queued is never really discarded (except on events like seek resets/reconfig/VO destruction). Even if a frame is dropped, it's assigned to `in->current_frame`. That's because if nothing is actually drawn after a dropped frame, the dropped frame must be rendered anyway. `do_redraw()` has logic for this. The semantic is actually that once the player core queued a frame, it's logically present until the next frame is queued, _especially_ if no next frame is queued.

So, considering all these messy circumstances, I think it'd be better if this frame dropping happened in the same place normal frame dropping happens... somehow.